### PR TITLE
Byte Implementation: Correct usage of size parameter for Rs256Decoder/Encoder

### PR DESCRIPTION
**Pull request by sonatique**
and especially he fact is was used instead of 256 for the size of the GaloisField

### CODE REVIEW

**Comment by antiduh** [comment]
I'm sorry, but I think I'm going to reject this pull request because your changes break the intended usage of the Rs256 classes due to an apparent misunderstanding of the field and the Rs256 classes.

The Rs256 classes can be used to store fields of _up to_ 256 elements (8-bits), but it certainly can support smaller fields.  The example cases you modified used fields of 16 elements (4-bit fields) and you broke them by changing the size of the underlying field.

For example, the regular (int-based) encoder theoretically supports fields of 4 billion elements (32-bit fields) because it uses C# ints to store elements; that doesn't mean we should hard code its size to 2^32.  

Both implementations can be used for fields of multiple sizes. Both implementations have theoretical and practical limits on the size of the fields they can support.

If you want to use Rs256* to operate a 256-element field, you can do so without modifying any of the code; just change which parameters you pass in. There's nothing different between Rs256Encoder and Encoder, except I copied Encoder to Rs256Encoder and changed all ints to bytes.

I'll leave the pull request open to receive you feedback, but I don't think I'll be merging your changes.

**Comment by sonatique** [comment]
Hello,
Thanks a lot for having a look at my code and taking the time for replying to me.

Let me provide some background info about what I did and why, but first let me thank you really very much for making your code freely available to everyone, it really helped me a lot.

As you correctly guessed I do not have the mathematical knowledge required to easily implement Reed-Solomon code by myself from scratch, but I have been able to use yours to get the building blocks I needed.

So: for some reason I had to implement a “(64, 60) Reed-Solomon code defined over GF(2^8)”, as specified. I have reference data for it, such that I can test whether a given implementation gives the expected results.

In this context (64, 60) means that we have a “message” of length 60, a “codeword” of total length 64, and thus a “parity section” of length 4. 64 and 60 are numbers of element whereas 2^8=256 is the number of possible state each of these element can take. So, my understanding is that this is a way to say we are operating on regular byte sequences. And this is of course confirmed by the rest of the context and by the reference data: we have 60 byte messages we want to complement with 4 byte parity words in order to form 64 bytes FEC encoded block to be sent.

Then I found your repository and in particular, I found the “ByteImpl” folder. Rs256Encoder/Decoder appeared to be exactly what I needed, so I started playing around with my reference data.

I first tried by constructing Rs256Encoder(64, 60, 0x11D) as “size” parameters seemed to be the total size of the codeword and “decodedSize” the size of the message.

It didn’t work out of the box for 2 reasons. For making your code spit out the same values than my reference data it needed some work:

1) I had to specify 65 instead of 64 for the size parameter, such that “Checkwords”, “EncodedSize” and “DecodedSize” can match 4, 64, 60 respectively. I was puzzled about why, and then by looking at your implementation I realized that “size” parameter was actually always used as “size -1”. So basically 65 was never used, 64 was use instead internally everywhere. There might be a mathematical reason for asking the user to say 65 when it means 64, but it I didn’t find it, and in general I prefer to not have to do minus 1 everywhere in my code. For me it seemed to make more sense to pass 64, 60 for telling the code to work as the (64, 60) RS defined in my problem.

2) Even with 65, there was something wrong (i.e. “IndexOutOfRangeException” in the calculations of GaloisField256.BuildLogarithms). And I found out that when passing 65 as the length of the size in the constructor of Rs256Encoder, the same value size = 65 was passed to the size parameter of the constructor of the GaloisField. 

This I don’t understand: why should the element count of the codeword and the size of the GaloisField be equal? What if I would like a codeword of size 64 and a GaloisField of size 256 or any other combination?

And actualy by the way: GaloisField256.BuildLogarithms throws an IndexOutOfRangeException for all values in between 2 and 256 except for 2, 3, 16, 256.
Again I don’t know the reason behind it, but I think this is an issue from a programmatic point of view.

So well, in the end: my understanding was that:
- We could simplify the code by removing all the “-1” and simply specify a value parameter of the useful value (pass x and use x, instead of passing x and use x-1)
- The GaloisField size should not necessarily be the codeword element count

So I made both changes. But yes: I did not realize that your Unit Tests were actually defined over less than 256 for the Galois Field size, so yes I broke it, that’s bad

So obviously my PR is plain wrong, but still, I wonder:

What does the “256” in Rs256Encode/Decoder and GaloisField256 stands for, then?

Why is codeword length passed directly as the GaloisField size? Shouldn’t you allow another parameter for the GaloisField size to be passed to the Encode/Decoder constructor?

Why is there such an equality of codeword length + 1 and GaloisField size in all of your unit tests?

Well: you don’t need to answer to me into much details, it is up to you. I am probably completely off-topic here, so you may not want to waste your time with me.

Anyway: I was able to achieve exactly what I needed by building on your code and those changes, and for this I thank you very much again.

Best regards.

**Comment by antiduh** [comment]
There's a deep, mathematical connection between the size of the Galois field and block sizes, and you see those invariants all throughout the source code.

RS encodes in blocks, and the size of each encoded block is one less than the number of elements in the field. That's it, there's no getting around it, it's how the polynomial math of RS works.

Consider GF(2^4). 

* Your symbols are base 2 (binary)
* Each symbol is 4 bits.
* The size of the field is 16 total symbols.
* The size of an encoded block is 15 symbols.
* Thus each block represents 15 * 4 bits == 60 bits.

Thus, if you want to encode and decode messages in GF(2^4), you have to provide it a byte buffer that has 15 places in it. It'll only use the bottom 4 bits of each place (each place can only be values 0-15, because those are the values of the field). You tell it how to divide the buffer into your original message values and parity values that it calculates and inserts for you.

If you were using the int-based implementation, you give it an 15-element array of ints and it does the same thing - values must be limited to 0-15, it only uses the bottom 4 bits of each element in the array, etc.

Using GF(2^4) is awkward for a software developer because you're working 4 bits at a time. Alright, so most people operate on GF(2^8) because it uses whole bytes.

Alright, so hows the block size math work out for GF(2^8)?

* Your symbols are still base 2 (binary).
* Each symbol is 8 bits (handy, because that's a byte).
* The size of the field is 256 symbols (aka, the number of unique values a byte can have).
* The size of an encoded block is 255 symbols.
* Thus each block represents 255 bytes, or 255 * 8 = 2040 bits

If you're using Rs256Encoder, you have to give it a byte array of 255 elements. If you're using Encoder (the int-based one), you have to give it an int array of 255 elements.

What if you want your message sizes to be smaller? Well, for the parts of the block that are yours to fill in ("original message symbols"), you're free to put whatever values you want there. When you give the block to the encoder, it'll fill in the parts of the block for the parity. 

So if you can give it whatever values you want, why not just give it zeros for the places you don't want to use?

 A (64, 60) RS code over GF(2^8) leaves 255-64 == 191 array elements set to zero. 60 contain original message values, 4 are left for the encoder to fill in with the parity values.

It may be helpful to read the big comment on this class:
https://github.com/antiduh/ErrorCorrection/blob/master/ErrorCorrection/GaloisField.cs

**Comment by antiduh** [comment]
Oh and to say it explicitly: 

If your field size is 2^8 or less, then Rs256Encoder and Encoder are interchangable. One operates on arrays of bytes, the other operates on arrays of ints. Encoder tends to have better performance, Rs256Encoder tends to be easier to work with when using bytes.

The 256 in Rs256 just means that it can only be used for fields of a maximum of 256 total elements.

**Comment by sonatique** [comment]
Hello again,

Thanks a lot for your explanation.

I still don't understand how to use your code to make my RS(64,60) operating on full 256 value bytes without modifications, given constructing Rs256Encoder(65, 60, poly) throws an exception in GaloisField.BuildLogarithm.

Shall I use Rs256Encoder(256, 60, poly) and just waste 191 bytes in my arrays (this means I'll have to to copy, allocation, etc.).

Why not just modify your code and let me specify Rs256Encoder(64, 60, 256, poly) or something similar? It will work straight without exception and without wasting byte in arrays... Considering your explanation, Rs256Encoder(int codewordSize, int messageSize, int galoisFieldSize, int poly) would make sens for all codewordSize <= galoisFieldSize), right?

And why does GaloisField throw for anything else than 2, 3, 16, 256? I see why 2, 16, 256 make sense (it's 2^1, 2^4, 2^8) but what about 3? And what about field of size 2^2, 2^3, etc.?

Thanks.

**Comment by antiduh** [comment]
> RS(64,60) operating on full 256 value bytes

Are you absolutely sure that someone wants you to implement a (64,60) code over GF(2^8)? What makes you think you need this? Can you provide a reference that I can read?

> Shall I use Rs256Encoder(256, 60, poly) and just waste 191 bytes in my arrays (this means I'll have to to copy, allocation, etc.).

It has to be done somewhere. Either you do it when passing the array to the encoders and decoders, or the encoders and decoders would have to do it for you because they have to run the complete algorithm across all 255 positions. It's the same amount of work no matter how it's implemented. 

> Why not just modify your code and let me

I'm not preventing you from doing anything except modifying my repository. My source code is open, it's BSD licensed. You can copy the source code into your project and modify it however you want. I don't care and my license says I don't care. The only thing you need from me is help understanding how this technology works, which I'm happy to provide (.. to a limit).

I feel like you didn't read that giant comment I pointed you to, because it sets the foundation for how a lot of this works and would answer your questions. I'm not willing to do the work for you if you're not willing to do the work for yourself.

> And why does GaloisField throw for anything else than 2, 3, 16, 256? I see why 2, 16, 256 make sense (it's 2^1, 2^4, 2^8) but what about 3? And what about field of size 2^2, 2^3, etc.?

Are you providing a valid field generator polynomial for the given set of parameters? I would not expect the same poly to work on different fields. Feel free to use the PrimeFinder method in IntImplTester to find valid primitive elements to use as field generator polynomials.

**Comment by sonatique** [comment]
Hello,
Thanks again. 
Sorry if I gave you the feeling I wanted you to do work for me. I don't need anything on my side, given I was able to implement exactly what I wanted by building upon your code. I can calculate exactly what I want, even without wasting byte in an array, everything is fine. Thanks again for your source code.

My proposal was just that you could, if you like, improve your code (from my point of view) by letting the user specify the size of the Galois Field independently of the length of the codeword, since I can see practical use for it. The code change would be easy.

Yes, I am absolutely sure that someone needs the (64, 60) RS code on GF(256). But unfortunately I cannot share anything, sorry.

However, I guess that many practical application of RS code operate on GF(256). For instance I found here: https://www.sciencedirect.com/topics/engineering/reed-solomon-code that Compact Disk use (32, 28) and (28, 24) RS Code over GF(256): "_The audio compact disk (CD) employs two Reed–Solomon codes in a cross-interleaved format to form a cross-interleaved RS code (CIRC). Each code operates over GF (2^8), therefore there are 8-bits per symbol. The first is a (32,28) code and the second a (28,24) code; each is double error correcting. The first code, therefore, has codewords each 32 symbols long, comprising 28 data and 4 parity. In the second code, codewords are formed from 24 data symbols and 4 parity symbols_." 

I think that when codeword length is smaller than GF size, it is called a **shortened** RS code, that's maybe what I should have said from the beginning. My (64, 60) is a **shortened** RS code operating on full byte / GF(256). 
There are many explanations and practical examples of such shortened codes:
https://ipnpr.jpl.nasa.gov/progress_report/42-75/75B.PDF, https://www.cs.cmu.edu/~guyb/realworld/reedsolomon/reed_solomon_codes.html
https://ieeexplore.ieee.org/document/7016053
etc.
The wikipedia page on RS code also mention shortened codes. 

So to summarize: I think that your Rs256Encoder/Decoder could benefit from a third parameter, the GF size, such that they became "Shortened code -compatible", for free, basically.

Anyways, thanks again for your time!
Best regards.

**Comment by antiduh** [comment]
If you have test vectors for the shortened codes, I wouldn't mind giving it a shot. This project could use some professional love after these these years.

**Comment by sonatique** [comment]
You can have a look here: https://github.com/sonatique/ErrorCorrection/blob/master/UnitTests/TestRs.cs

I quickly put 4 Unit Tests, the code is a bit rough, but tests are valid
